src/systems/filecoin_vm/interpreter/vm_interpreter.go

```diff
@@ -14,172 +15,255 @@ import (
+	// use ControlActor instead (https://github.com/filecoin-project/specs/issues/665)
+
+	firstMiner := msgs.Blocks()[0].Miner()
+	cronMessage := _makeCronTickMessage(outTree, firstMiner)
+	outTree = _applyMessageBuiltinAssert(outTree, cronMessage, firstMiner)
 
 	return
 }
 
-func (vmi *VMInterpreter_I) ApplyMessage(inTree st.StateTree, message msg.UnsignedMessage, minerAddr addr.Address) (
-	st.StateTree, vmr.MessageReceipt) {
+func (vmi *VMInterpreter_I) ApplyMessage(
+	inTree st.StateTree, message msg.UnsignedMessage, onChainMessageSize int, minerAddr addr.Address) (
+	retTree st.StateTree, retReceipt vmr.MessageReceipt, retMinerPenalty actor.TokenAmount) {
 
-	compTree := inTree
-	var outTree st.StateTree
-	var toActor actor.ActorState
-	var err error
+	vmiGasRemaining := message.GasLimit()
+	vmiGasUsed := msg.GasAmount_Zero()
+
+	_applyReturn := func(
+		tree st.StateTree, invocOutput vmr.InvocOutput, exitCode exitcode.ExitCode,
+		senderResolveSpec SenderResolveSpec) {
+
+		switch senderResolveSpec {
+		case SenderResolveSpec_OK:
+			// In this case, the sender is valid and has already transferred funds to the burnt funds actor
+			// sufficient for the gas limit. Thus, we may refund the unused gas funds to the sender here.
+			Assert(!message.GasLimit().LessThan(vmiGasUsed))
+			Assert(message.GasLimit().Equals(vmiGasUsed.Add(vmiGasRemaining)))
+			vmiGasRemainingFIL := _gasToFIL(vmiGasRemaining, message.GasPrice())
+			tree = _withTransferFundsAssert(tree, addr.BurntFundsActorAddr, message.From(), vmiGasRemainingFIL)
+			retMinerPenalty = actor.TokenAmount(0)
+
+		case SenderResolveSpec_Invalid:
+			vmiGasUsedFIL := _gasToFIL(vmiGasUsed, message.GasPrice())
+			retMinerPenalty = vmiGasUsedFIL
 
-	fromActor := compTree.GetActorState(message.From())
+		default:
+			Assert(false)
+		}
+
+		retTree = tree
+		retReceipt = vmr.MessageReceipt_Make(invocOutput, exitCode, vmiGasUsed)
+	}
+
+	_applyError := func(tree st.StateTree, errExitCode exitcode.SystemErrorCode, senderResolveSpec SenderResolveSpec) {
+		_applyReturn(tree, vmr.InvocOutput_Make(nil), exitcode.SystemError(errExitCode), senderResolveSpec)
+	}
+
+	// Deduct an amount of gas corresponding to cost about to be incurred, but not necessarily
+	// incurred yet.
+	_vmiAllocGas := func(amount msg.GasAmount) (vmiAllocGasOK bool) {
+		vmiGasRemaining, vmiAllocGasOK = vmiGasRemaining.SubtractIfNonnegative(amount)
+		vmiGasUsed = message.GasLimit().Subtract(vmiGasRemaining)
+		Assert(!vmiGasRemaining.LessThan(msg.GasAmount_Zero()))
+		Assert(!vmiGasUsed.LessThan(msg.GasAmount_Zero()))
+		return
+	}
+
+	// Deduct an amount of gas corresponding to costs already incurred, and for which the
+	// gas cost must be paid even if it would cause the gas used to exceed the limit.
+	_vmiBurnGas := func(amount msg.GasAmount) (vmiBurnGasOK bool) {
+		vmiGasUsedPre := vmiGasUsed
+		vmiBurnGasOK = _vmiAllocGas(amount)
+		if !vmiBurnGasOK {
+			Assert(vmiGasRemaining.Equals(msg.GasAmount_Zero()))
```

Comment:
That's not how `SubtractIfNonnegative` works. `vmiGasRemaining` will be its prior value, not zero.